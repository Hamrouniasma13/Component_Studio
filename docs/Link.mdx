---
name: Link
route: /link
menu: ðŸ§° Technical
---

import { Playground } from 'docz'
import { Link } from '../src'

# Link

`Link` is the main component component when it comes to make a link between 2 pages.
It comes with no styles, but by default, it resets the color and the text-decoration.

<Playground>
  <Link href="/">Link to home</Link>
</Playground>

## Customize

Be default, we use the html `a` tag, but we can use the element of your choice

_PS: As our `Button` and `Simple Button` use `Link`, you should not use them as custom element, or it will create a infinite loop_

<Playground>
  <Link.ContextProvider element={props => <button {...props} />}>
    <Link href="/">Call to action</Link>
  </Link.ContextProvider>
</Playground>

In the example above, we use `Link.ContextProvider` to put a `button` instead of the default `a` when we make a link.
You can also use other tags, other elements from other libraries, or modify the props, like `onClick`.
