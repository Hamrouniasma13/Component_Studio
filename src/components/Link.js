/** @jsx jsx */
import { jsx } from '@emotion/core'
import React from 'react'
import PropTypes from 'prop-types'

/**
 * LinkContext allow users of this library to use the component they want as a link. By defaut we use 'a', but
 * they could use 'button', or Link from react-router, or add some function on the click, like analytics tracking.
 */
const LinkContext = React.createContext('a')
const LinkProvider = ({ element = 'a', children }) => (
  <LinkContext.Provider value={element}>{children}</LinkContext.Provider>
)

LinkProvider.propTypes = {
  element: PropTypes.any
}

export default function Link({ css, ...props }) {
  // Retrieve the custom html tag we'll use
  const Element = React.useContext(LinkContext)

  // Remove browser default style
  const resetStyle = { color: 'inherit', textDecoration: 'none' }

  return <Element css={{ ...resetStyle, ...css }} {...props} />
}

Link.ContextProvider = LinkProvider
