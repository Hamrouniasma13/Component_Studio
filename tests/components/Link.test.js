import React from 'react';
import renderer from 'react-test-renderer';

// wrapper to make our components work with <Link />
import { MemoryRouter } from 'react-router';

import { render, mount, shallow } from 'enzyme';

import Link from '../../components/Link';

describe('Component: <Link />', () => {
  it('render snapshot', () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={['/']}>
          <Link href="/">View More</Link>
        </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('href to external url', () => {
    const wrapper = render(<Link href="http://google.com">Go to google</Link>);

    expect(wrapper.attr('href')).toBe('http://google.com');
  });

  it('trigger onClick() callback', () => {
    const onButtonClick = jest.fn();
    const wrapper = mount(
      <Link className="my-link" onClick={onButtonClick}>
        Open a new page
      </Link>
    );
    // console.log('wrapper: ', wrapper.debug());

    wrapper.find('a.my-link').simulate('click');

    expect(onButtonClick).toBeCalled();
  });

  it('render with "to" object', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <Link
          href={{
            path: '/login',
            search: '?utm=your+face',
            state: { referrer: 'about-us' }
          }}
        />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });
});
