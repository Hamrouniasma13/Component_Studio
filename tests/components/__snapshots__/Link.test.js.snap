// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Link /> render snapshot 1`] = `
<a
  aria-current="page"
  className="active"
  href="/"
  onClick={[Function]}
  style={Object {}}
>
  View More
</a>
`;

exports[`Component: <Link /> render with "to" object 1`] = `
<a
  aria-current="page"
  className="active"
  href="/?utm=your+face"
  onClick={[Function]}
  style={Object {}}
/>
`;
